File: services/dataService.js
const fs = require('fs').promises;
const path = require('path');

class DataService {
  constructor() {
    this.dataDir = path.join(__dirname, '../data');
    this.applicationsFile = path.join(this.dataDir, 'applications.json');
    this.teamsFile = path.join(this.dataDir, 'teams.json');
    this.initializeDataFiles();
  }

  async initializeDataFiles() {
    try {
      // Create data directory if it doesn't exist
      await fs.mkdir(this.dataDir, { recursive: true });

      // Initialize applications file
      try {
        await fs.access(this.applicationsFile);
      } catch {
        await fs.writeFile(this.applicationsFile, JSON.stringify([], null, 2));
      }

      // Initialize teams file with default teams
      try {
        await fs.access(this.teamsFile);
      } catch {
        const defaultTeams = [
          {
            id: 'engineering',
            name: 'Engineering',
            description: 'Software development and technical roles',
            executives: ['cto@company.com'],
            projectLeads: ['eng-lead@company.com'],
            createdAt: new Date().toISOString()
          },
          {
            id: 'product',
            name: 'Product',
            description: 'Product management and design roles',
            executives: ['cpo@company.com'],
            projectLeads: ['product-lead@company.com'],
            createdAt: new Date().toISOString()
          },
          {
            id: 'marketing',
            name: 'Marketing',
            description: 'Marketing and growth roles',
            executives: ['cmo@company.com'],
            projectLeads: ['marketing-lead@company.com'],
            createdAt: new Date().toISOString()
          }
        ];
        await fs.writeFile(this.teamsFile, JSON.stringify(defaultTeams, null, 2));
      }

      console.log('✅ Data service initialized successfully');
    } catch (error) {
      console.error('❌ Failed to initialize data service:', error.message);
    }
  }

  // Applications methods
  async saveApplication(application) {
    try {
      const applications = await this.getAllApplications();
      applications.push(application.toJSON());
      await fs.writeFile(this.applicationsFile, JSON.stringify(applications, null, 2));
      return application;
    } catch (error) {
      throw new Error(`Failed to save application: ${error.message}`);
    }
  }

  async getAllApplications() {
    try {
      const data = await fs.readFile(this.applicationsFile, 'utf8');
      return JSON.parse(data);
    } catch (error) {
      console.error('Error reading applications:', error.message);
      return [];
    }
  }

  async getApplicationById(id) {
    try {
      const applications = await this.getAllApplications();
      return applications.find(app => app.id === id);
    } catch (error) {
      throw new Error(`Failed to get application: ${error.message}`);
    }
  }

  async updateApplicationStatus(id, status) {
    try {
      const applications = await this.getAllApplications();
      const appIndex = applications.findIndex(app => app.id === id);
      
      if (appIndex === -1) {
        throw new Error('Application not found');
      }

      applications[appIndex].status = status;
      await fs.writeFile(this.applicationsFile, JSON.stringify(applications, null, 2));
      
      return applications[appIndex];
    } catch (error) {
      throw new Error(`Failed to update application status: ${error.message}`);
    }
  }

  // Teams methods
  async saveTeam(team) {
    try {
      const teams = await this.getAllTeams();
      const existingIndex = teams.findIndex(t => t.id === team.id);
      
      if (existingIndex !== -1) {
        teams[existingIndex] = team.toJSON();
      } else {
        teams.push(team.toJSON());
      }
      
      await fs.writeFile(this.teamsFile, JSON.stringify(teams, null, 2));
      return team;
    } catch (error) {
      throw new Error(`Failed to save team: ${error.message}`);
    }
  }

  async getAllTeams() {
    try {
      const data = await fs.readFile(this.teamsFile, 'utf8');
      return JSON.parse(data);
    } catch (error) {
      console.error('Error reading teams:', error.message);
      return [];
    }
  }

  async getTeamById(id) {
    try {
      const teams = await this.getAllTeams();
      return teams.find(team => team.id === id);
    } catch (error) {
      throw new Error(`Failed to get team: ${error.message}`);
    }
  }

  async getTeamByName(name) {
    try {
      const teams = await this.getAllTeams();
      return teams.find(team => team.name.toLowerCase() === name.toLowerCase());
    } catch (error) {
      throw new Error(`Failed to get team by name: ${error.message}`);
    }
  }

  async deleteTeam(id) {
    try {
      const teams = await this.getAllTeams();
      const filteredTeams = teams.filter(team => team.id !== id);
      
      if (teams.length === filteredTeams.length) {
        throw new Error('Team not found');
      }
      
      await fs.writeFile(this.teamsFile, JSON.stringify(filteredTeams, null, 2));
      return true;
    } catch (error) {
      throw new Error(`Failed to delete team: ${error.message}`);
    }
  }
}

module.exports = new DataService();

File: src/main/java/com/sfuosdev/emailbackend/service/DataService.java
package com.sfuosdev.emailbackend.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.sfuosdev.emailbackend.model.Application;
import com.sfuosdev.emailbackend.model.Team;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class DataService {
    private static final Logger logger = LoggerFactory.getLogger(DataService.class);
    private final ObjectMapper objectMapper;
    private final Path dataDir;
    private final Path applicationsFile;
    private final Path teamsFile;
    
    public DataService() {
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        this.dataDir = Paths.get("data");
        this.applicationsFile = dataDir.resolve("applications.json");
        this.teamsFile = dataDir.resolve("teams.json");
        initializeDataFiles();
    }
    
    private void initializeDataFiles() {
        try {
            // Create data directory if it doesn't exist
            Files.createDirectories(dataDir);
            
            // Initialize applications file
            if (!Files.exists(applicationsFile)) {
                saveApplications(new ArrayList<>());
            }
            
            // Initialize teams file with default teams
            if (!Files.exists(teamsFile)) {
                List<Team> defaultTeams = createDefaultTeams();
                saveTeams(defaultTeams);
            }
            
            logger.info("✅ Data service initialized successfully");
        } catch (Exception e) {
            logger.error("❌ Failed to initialize data service: {}", e.getMessage());
        }
    }
    
    private List<Team> createDefaultTeams() {
        List<Team> teams = new ArrayList<>();
        
        Team engineering = new Team("Engineering", "Software development and technical roles");
        engineering.setId("engineering");
        engineering.getExecutives().add("cto@company.com");
        engineering.getProjectLeads().add("eng-lead@company.com");
        teams.add(engineering);
        
        Team product = new Team("Product", "Product management and design roles");
        product.setId("product");
        product.getExecutives().add("cpo@company.com");
        product.getProjectLeads().add("product-lead@company.com");
        teams.add(product);
        
        Team marketing = new Team("Marketing", "Marketing and growth roles");
        marketing.setId("marketing");
        marketing.getExecutives().add("cmo@company.com");
        marketing.getProjectLeads().add("marketing-lead@company.com");
        teams.add(marketing);
        
        return teams;
    }
    
    // Application methods
    public Application saveApplication(Application application) throws IOException {
        List<Application> applications = getAllApplications();
        applications.add(application);
        saveApplications(applications);
        return application;
    }
    
    public List<Application> getAllApplications() {
        try {
            if (!Files.exists(applicationsFile)) {
                return new ArrayList<>();
            }
            return objectMapper.readValue(applicationsFile.toFile(), new TypeReference<List<Application>>() {});
        } catch (IOException e) {
            logger.error("Error reading applications: {}", e.getMessage());
            return new ArrayList<>();
        }
    }
    
    public Optional<Application> getApplicationById(String id) {
        return getAllApplications().stream()
                .filter(app -> app.getId().equals(id))
                .findFirst();
    }
    
    public Application updateApplicationStatus(String id, String status) throws IOException {
        List<Application> applications = getAllApplications();
        Optional<Application> applicationOpt = applications.stream()
                .filter(app -> app.getId().equals(id))
                .findFirst();
        
        if (applicationOpt.isEmpty()) {
            throw new IllegalArgumentException("Application not found");
        }
        
        Application application = applicationOpt.get();
        application.setStatus(status);
        saveApplications(applications);
        return application;
    }
    
    private void saveApplications(List<Application> applications) throws IOException {
        objectMapper.writerWithDefaultPrettyPrinter().writeValue(applicationsFile.toFile(), applications);
    }
    
    // Teams methods
    public Team saveTeam(Team team) throws IOException {
        List<Team> teams = getAllTeams();
        teams.removeIf(t -> t.getId().equals(team.getId()));
        teams.add(team);
        saveTeams(teams);
        return team;
    }
    
    public List<Team> getAllTeams() {
        try {
            if (!Files.exists(teamsFile)) {
                return new ArrayList<>();
            }
            return objectMapper.readValue(teamsFile.toFile(), new TypeReference<List<Team>>() {});
        } catch (IOException e) {
            logger.error("Error reading teams: {}", e.getMessage());
            return new ArrayList<>();
        }
    }
    
    public Optional<Team> getTeamById(String id) {
        return getAllTeams().stream()
                .filter(team -> team.getId().equals(id))
                .findFirst();
    }
    
    public Optional<Team> getTeamByName(String name) {
        return getAllTeams().stream()
                .filter(team -> team.getName().equalsIgnoreCase(name))
                .findFirst();
    }
    
    public void deleteTeam(String id) throws IOException {
        List<Team> teams = getAllTeams();
        boolean removed = teams.removeIf(team -> team.getId().equals(id));
        if (!removed) {
            throw new IllegalArgumentException("Team not found");
        }
        saveTeams(teams);
    }
    
    private void saveTeams(List<Team> teams) throws IOException {
        objectMapper.writerWithDefaultPrettyPrinter().writeValue(teamsFile.toFile(), teams);
    }
    
    // Statistics methods
    public List<Application> getRecentApplications(int days) {
        LocalDateTime cutoff = LocalDateTime.now().minusDays(days);
        return getAllApplications().stream()
                .filter(app -> app.getAppliedAt().isAfter(cutoff))
                .toList();
    }
}